028-service-binding.md

```diff
@@ -0,0 +1,374 @@
+
+The Service Binding specification defines how connection information (e.g. endpoints, username, password) for services such as databases and message brokers are made available to runtime applications in Kubernetes. Version 1 of the specification is due to be released shortly.
+
+Today, Strimzi does not fit very nicely with service binding, it both lacks the status field required by the Service Binding operator and does not make credentials available in a single Secret which is what the Service Binding Operator expects from services that implement the spec.
+
+Contents:
+
+ - [Current situation](#current-situation)
+ - [Proposal](#proposal)
+ - [Rejected Alternatives](#rejected-alternatives)
+ - [Overview of Service Binding specification](#overview-of-service-binding-specification)
+ - [Connecting to Strimzi today](#connecting-to-strimzi-today)
+
+## Current situation
+
+At a high level these are the current problems that make integrating Strimzi with a Service Binding Operator hard:
+
+1. Bootstrap Server information is not contained in any secret
+2. Information is spread across multiple `Secret` resources
+3. The user has to determine which `KafkaUser` can access which listener
+
+## Proposal
+
+This proposal introduces a new custom resource type called `KafkaConnection` and the application creates a `ServiceBinding` that binds to this service.
+
+Required changes:
+
+ - Add a new Custom Resource called `KafkaConnection` that refers to the `Kafka`, the specific listener and the `KafkaUser` that should be used
+ - The Strimzi operator adds a `status.binding` to the `KafkaConnection` status that points to a Kubernetes `Secret`
+ - The Strimzi operator creates a `Secret` in the same namespace as the `KafkaConnection` containing all the required information
+
+`KafkaConnection` CR spec:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnection
+metadata:
+  name: barista-kafka
+spec:
+  kafka:
+    apiVersion: kafka.strimzi.io/v1beta2
+    kind: Kafka
+    name: my-cluster
+    namespace: my-namespace # optional
+    listener: # optional
+      name: tls # optional
+  user: # optional
+    apiVersion: kafka.strimzi.io/v1beta2
+    kind: KafkaUser
+    name: my-barista
+    namespace: my-namespace # optional
+status:
+  binding:
+    name: barista-kafka
+```
+
+`Secret` created by Strimzi:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: barista-kafka
+  namespace: my-namespace
+type: Opaque
+  data:
+    type: kafka
+    provider: strimzi
+    bootstrapServer: # comma separated list of host:port
```

Comment:
Maybe worth mentioning that we use `bootstrapServer` rather than the more standard `host` and `port` because most Kafka clients are configured with a list of bootstrap servers, so using `host` and `port` would be more inconvenient.